src/legacy/core_plugins/kibana/public/dev_tools/render_app.tsx

```diff
@@ -0,0 +1,180 @@
+  unmountHandler: () => void;
+}
+
+function DevToolsWrapper({
+  devTools,
+  activeDevTool,
+  appMountContext,
+  updateRoute,
+}: DevToolsWrapperProps) {
+  const mountedTool = useRef<MountedDevToolDescriptor | null>(null);
+
+  useEffect(
+    () => () => {
+      if (mountedTool.current) {
+        mountedTool.current.unmountHandler();
+      }
+    },
+    []
+  );
+
+  return (
+    <main className="devApp">
+      <EuiTabs>
+        {devTools.map(currentDevTool => (
+          <EuiToolTip content={currentDevTool.tooltipContent} key={currentDevTool.id}>
+            <EuiTab
+              disabled={currentDevTool.disabled}
+              isSelected={currentDevTool === activeDevTool}
+              onClick={() => {
+                if (!currentDevTool.disabled) {
+                  updateRoute(`/dev_tools/${currentDevTool.id}`);
+                }
+              }}
+            >
+              {currentDevTool.title}
+            </EuiTab>
+          </EuiToolTip>
+        ))}
+      </EuiTabs>
+      <div
+        className="devApp__container"
+        role="tabpanel"
+        data-test-subj={activeDevTool.id}
+        ref={async element => {
+          if (
+            element &&
+            (mountedTool.current === null ||
+              mountedTool.current.devTool !== activeDevTool ||
+              mountedTool.current.mountpoint !== element)
+          ) {
+            if (mountedTool.current) {
+              mountedTool.current.unmountHandler();
+            }
+            const unmountHandler = await activeDevTool.mount(appMountContext, {
+              element,
+              appBasePath: '',
+            });
+            mountedTool.current = {
+              devTool: activeDevTool,
+              mountpoint: element,
+              unmountHandler,
+            };
+          }
+        }}
+      />
+    </main>
+  );
+}
+
+function redirectOnMissingCapabilities(appMountContext: AppMountContext) {
+  if (!appMountContext.core.application.capabilities.dev_tools.show) {
+    window.location.hash = '/home';
+  }
+}
+
+function setBadge(appMountContext: AppMountContext) {
+  if (appMountContext.core.application.capabilities.dev_tools.save) {
+    return;
+  }
+  appMountContext.core.chrome.setBadge({
+    text: i18n.translate('kbn.devTools.badge.readOnly.text', {
+      defaultMessage: 'Read only',
+    }),
+    tooltip: i18n.translate('kbn.devTools.badge.readOnly.tooltip', {
+      defaultMessage: 'Unable to save',
+    }),
+    iconType: 'glasses',
+  });
+}
+
+function setBreadcrumbs(appMountContext: AppMountContext) {
+  appMountContext.core.chrome.setBreadcrumbs([
+    {
+      text: i18n.translate('kbn.devTools.k7BreadcrumbsDevToolsLabel', {
+        defaultMessage: 'Dev Tools',
+      }),
+      href: '#/dev_tools',
+    },
+  ]);
+}
+
+export function renderApp(
+  element: HTMLElement,
+  appMountContext: AppMountContext,
+  basePath: string,
+  devTools: DevTool[]
+) {
+  redirectOnMissingCapabilities(appMountContext);
```

Comment:
Not sure whether this matters, but could we explicitly stop the flow of execution here? Or is setting `window.location.hash` sufficient for this?

Reply:
Good point, it's probably cleaner to not render at all. Changed.